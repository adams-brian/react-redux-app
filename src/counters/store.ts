import { IAction } from '../store';

export interface IState {
  counters: number[];
}

/* tslint:disable:max-classes-per-file */
abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
export type CountersAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const actionCreators = {
  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),
};

function countersReducer(state: number[] = [], action: CountersAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

export const reducers = {
  counters: countersReducer
}
