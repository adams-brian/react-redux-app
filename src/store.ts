import { LocationChangeAction } from 'react-router-redux';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IAppState {
  loading: boolean;
  counters: number[];
  users: IUser[];
}

interface IAction {
  type: string;
}

/* tslint:disable:max-classes-per-file */
export const START_LOADING = '[App] START LOADING';
class StartLoading implements IAction {
  public readonly type = START_LOADING;
}
export const DONE_LOADING = '[App] DONE LOADING';
class DoneLoading implements IAction {
  public readonly type = DONE_LOADING;
}
type LoadingAction = StartLoading | DoneLoading;

abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
type CounterAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements IAction {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) {}
}
export const UPDATE_USER = '[Users] UPDATE';
class UpdateUser implements IAction {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) {}
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements IAction {
  public readonly type = DELETE_USER;
  constructor(public payload: string) {}
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements IAction {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) {}
}
type UserAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated;

export const actionCreators = {
  doneLoading: () => Object.assign({}, new DoneLoading()),
  startLoading: () => Object.assign({}, new StartLoading()),

  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),

  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

function loadingReducer(state: boolean = false, action: LoadingAction) {
  switch (action.type) {
    case START_LOADING:
      return true;
    case DONE_LOADING:
      return false;
    default:
      return state;
  }
}

function countersReducer(state: number[] = [], action: CounterAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

const usersReducer = (state: IUser[] = [], action: UserAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const reducers = {
  counters: countersReducer,
  loading: loadingReducer,
  users: usersReducer
};

export type AppAction = LoadingAction | CounterAction | UserAction | LocationChangeAction;
