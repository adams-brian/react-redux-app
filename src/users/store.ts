import { Action, combineReducers } from 'redux';

export interface IUser {
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IState {
  users: {
    error: string;
    list: IUser[];
    loaded: boolean;
  }
}

/* tslint:disable:max-classes-per-file */
export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements Action {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) { }
}
export const UPDATE_USER = '[Users] UPDATE';
export class UpdateUser implements Action {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) { }
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements Action {
  public readonly type = DELETE_USER;
  constructor(public payload: string) { }
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements Action {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) { }
}
export const LOAD_USERS = '[Users] LOAD USERS';
export class LoadUsers implements Action {
  public readonly type = LOAD_USERS;
}
export const USERS_ERROR = '[Users] USERS ERROR';
export class UsersError implements Action {
  public readonly type = USERS_ERROR;
  constructor(public payload: Error) { }
}
export const USERS_LOADED = '[Users] USERS LOADED';
export class UsersLoaded implements Action {
  public readonly type = USERS_LOADED;
}
export type UsersAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated | LoadUsers | UsersError | UsersLoaded;

export const actionCreators = {
  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  loadUsers: () => Object.assign({}, new LoadUsers()),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersError: (err: Error) => Object.assign({}, new UsersError(err)),
  usersLoaded: () => Object.assign({}, new UsersLoaded()),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

export const listReducer = (state: IUser[] = [], action: UsersAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const loadedReducer = (state: boolean = false, action: UsersAction) => {
  switch (action.type) {
    case USERS_LOADED:
      return true;
    default:
      return state;
  }
}

export const errorReducer = (state: string = '', action: UsersAction) => {
  switch (action.type) {
    case USERS_ERROR:
      return action.payload.message;
    default:
      return state;
  }
}

export const reducers = {
  users: combineReducers({
    error: errorReducer,
    list: listReducer,
    loaded: loadedReducer
  })
};
