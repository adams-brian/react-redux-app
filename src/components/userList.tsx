import * as React from 'react';
import { Link } from 'react-router-dom';

import { IUser } from '../store';

interface IUserListProps {
  list: IUser[];
  deleteUser: (id: string) => void;
  editUser: (id: string) => void;
}

/* tslint:disable:jsx-no-lambda */
export const UserList = (props: IUserListProps) => (
  <div className="users-container">
    <h1 className="users-header">Users</h1>
    <table className="table table-hover">
      <thead>
        <tr>
          <th>First Name</th>
          <th>Last Name</th>
          <th style={{width: '1px'}}/>
        </tr>
      </thead>
      <tbody>
        {props.list.map(user => (
          <tr key={user._id} onClick={() => props.editUser(user._id)}>
            <td>{user.firstname}</td>
            <td>{user.lastname}</td>
            <td>
              <button onClick={(e) => { e.stopPropagation(); props.deleteUser(user._id); }}>
                <span className="fa fa-times"/>
              </button>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
    <Link className="create-user" to="/users/createuser">
      <button type="button" className="btn btn-primary">Create User</button>
    </Link>
  </div>
);
