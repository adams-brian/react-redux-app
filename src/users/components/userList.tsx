import { push } from 'connected-react-router';
import * as React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Action, Dispatch } from 'redux';

import { actionCreators, IState, IUser, UsersAction } from '../store';
import UserListRow from './userListRow';

interface IUserListProps {
  list: IUser[];
  deleteUser: (id: string) => void;
  editUser: (id: string) => void;
}

export const UserList = (props: IUserListProps) => (
  <div className="users-container">
    <h1 className="users-header">Users</h1>
    <table className="table table-hover">
      <thead>
        <tr>
          <th>First Name</th>
          <th>Last Name</th>
          <th style={{width: '1px'}}/>
        </tr>
      </thead>
      <tbody>
        {props.list.map(user =>
          <UserListRow 
            key={user._id}
            user={user}
            editUser={props.editUser}
            deleteUser={props.deleteUser}
          />
        )}
      </tbody>
    </table>
    <Link className="create-user" to="/users/createuser">
      <button type="button" className="btn btn-primary">Create User</button>
    </Link>
  </div>
);

export const mapStateToProps = (state: IState) => 
  ({ list: state.users.list });

export const mapDispatchToProps = (dispatch: Dispatch<UsersAction | Action<any>>) => ({
  deleteUser: (id: string) => dispatch(actionCreators.deleteUser(id)),
  editUser: (id: string) => { dispatch(push('/users/' + id)); }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)( UserList );
