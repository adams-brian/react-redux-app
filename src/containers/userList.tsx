import { connect } from 'react-redux';
import { AppState, actionCreators, AppAction } from '../store';

import { UserList as UnwrappedUserList } from '../components/userList';
import { Dispatch } from 'redux';
import { push } from 'react-router-redux';

export const UserList = connect(
  (state: AppState) => 
    ({ list: state.users }),
  (dispatch: Dispatch<AppAction>) => ({
    deleteUser: (id: string) => dispatch(actionCreators.deleteUser(id)),
    editUser: (id: string) => { dispatch(push('/users/' + id)); }
  })
  )(UnwrappedUserList);
