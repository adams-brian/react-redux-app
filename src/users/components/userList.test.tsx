import { shallow } from 'enzyme';
import * as React from 'react';
import { push } from 'react-router-redux';
import * as sinon from 'sinon';

import { actionCreators } from '../store';
import { mapDispatchToProps, mapStateToProps, UserList } from './userList';

type propFunction = (id: string) => void;

const list = [
  {
    _id: '123',
    firstname: 'abc',
    lastname: 'def'
  },
  {
    _id: '456',
    firstname: 'hij',
    lastname: 'klm'
  }
];

describe("UserList", () => {
  
  describe("component", () => {

    let element: JSX.Element;
    let deleteUser: sinon.SinonSpy;
    let editUser: sinon.SinonSpy;
    
    beforeEach(() => {
      
      deleteUser = sinon.spy();
      editUser = sinon.spy();

      element = (
        <UserList
          list={list}
          deleteUser={deleteUser}
          editUser={editUser}
        />
      );
    });

    it('renders as expected', () => {
      const component = shallow(element);
      expect(component).toMatchSnapshot();
    });

    it('calls deleteUser', () => {
      const component = shallow(element);
      expect(deleteUser.called).toBe(false);
      (component.find('UserListRow').at(0).prop('deleteUser') as propFunction)('123');
      expect(deleteUser.calledOnceWithExactly('123')).toBe(true);
    });

    it('calls editUser', () => {
      const component = shallow(element);
      expect(editUser.called).toBe(false);
      (component.find('UserListRow').at(1).prop('editUser') as propFunction)('456');
      expect(editUser.calledOnceWithExactly('456')).toBe(true);
    });

  });

  describe("connection", () => {

    it('maps state to props', () => {
      expect(mapStateToProps({ users: list })).toEqual({ list });
    });
  
    it('dispatches deleteUser', () => {
      const dispatch = sinon.spy();
      const props = mapDispatchToProps(dispatch);
      expect(dispatch.called).toBe(false);
      props.deleteUser('5');
      expect(dispatch.calledOnceWithExactly(actionCreators.deleteUser('5'))).toBe(true);
    });

    it('dispatches editUser', () => {
      const dispatch = sinon.spy();
      const props = mapDispatchToProps(dispatch);
      expect(dispatch.called).toBe(false);
      props.editUser('5');
      expect(dispatch.calledOnceWithExactly(push('/users/5'))).toBe(true);
    });

  });
  
});
