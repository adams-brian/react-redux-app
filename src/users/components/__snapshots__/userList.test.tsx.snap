// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList component renders as expected 1`] = `
<div
  className="users-container"
>
  <h1
    className="users-header"
  >
    Users
  </h1>
  <table
    className="table table-hover"
  >
    <thead>
      <tr>
        <th>
          First Name
        </th>
        <th>
          Last Name
        </th>
        <th
          style={
            Object {
              "width": "1px",
            }
          }
        />
      </tr>
    </thead>
    <tbody>
      <UserListRow
        deleteUser={[Function]}
        editUser={[Function]}
        key="123"
        user={
          Object {
            "_id": "123",
            "firstname": "abc",
            "lastname": "def",
          }
        }
      />
      <UserListRow
        deleteUser={[Function]}
        editUser={[Function]}
        key="456"
        user={
          Object {
            "_id": "456",
            "firstname": "hij",
            "lastname": "klm",
          }
        }
      />
    </tbody>
  </table>
  <Link
    className="create-user"
    replace={false}
    to="/users/createuser"
  >
    <button
      className="btn btn-primary"
      type="button"
    >
      Create User
    </button>
  </Link>
</div>
`;
